feat(broker-core): add debug exporter http server

menski opened the pull request:
- adds an optional http server to the debug exporter
- if debug exporter is configured with a port greater then zero the http server will be enabled
- the exporter then serves a html page of the latest records
- the records can be filtered by value type and full text search
- the maximum number of records served can be configured `httpLimit`

closes #1550

### Screenshot

![screenshot-2018-10-25-07 55 29](https://user-images.githubusercontent.com/516296/47478887-7ac93480-d82b-11e8-8dfc-836e0c289ea9.png)

saig0 commented on broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html:
```diff
@@ -0,0 +1,185 @@
+        return $('#filter-incident');
+      case 'TIMER':
+        return $('#filter-timer');
+      case 'MESSAGE_SUBSCRIPTION':
```
The value type `MESSAGE` should be also displayed if `Message` is selected.

menski replied:
Correct I forgot it, will change

saig0 commented on broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html:
```diff
@@ -0,0 +1,185 @@
+  </thead>
+  <tbody>
+  {{ #. }}
+    <tr data-toggle="collapse" data-target="#collapse-{{_id}}" aria-expanded="false" aria-controls="collapse-{{_id}}">
```
It should be more clear that I can click on the record to see the details. I found this feature by a fluke ;)

menski replied:
Any ideas how to visualize this? Maybe a plus icon in as first column?

saig0 replied:
Yeah, this would be an option. Or change the mouse symbol.

menski commented on broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporter.java:
```diff
@@ -37,6 +40,11 @@ public void configure(Context context) {
+
+    if (httpServer == null && configuration.httpPort > 0) {
+      httpServer =
+          new DebugExporterHttpServer(log, configuration.httpPort, configuration.httpLimit);
```
I wanted to start it only once, i.e. not for every partition, and verify that the port is available, which I think should happen in the initial configure call. But I can also move it to open if it is to strange.

saig0 replied:
Ok, make sense.

menski commented on broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  public synchronized void add(String recordJson) {
```
Cause multiple exporters, i.e. for every partition, will add records to the list.

saig0 replied:
You're right. I didn't see the static reference.

menski commented on broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java:
```diff
@@ -0,0 +1,123 @@
+    return null;
+  }
+
+  private byte[] loadIndexPage() {
```
Okay I can change that, but then I think I have to open the server on the initial configure call. At least I understood it that way. @npepinpe what do you think?

npepinpe replied:
If you fail in the configure call, then the broker will not start; if that's the intention then you'd have to throw an exception in the configure call (I'm still not 100% convinced it wouldn't have been better to have a separate `validate` call but that's another discussion)

menski commented on broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html:
```diff
@@ -0,0 +1,185 @@
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Zeebe Debug Exporter</title>
+  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
```
I didn't wanted to implement a static resource handler in the web server and keep it simple. But if you want I can also serve all resource with the HTTP server

menski commented on broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html:
```diff
@@ -0,0 +1,185 @@
+
+  function searchRecords(records) {
+    var searchTerm = $('#filter-search').val();
+    console.log(searchTerm);
```
oops forgot to remove the debugging code

menski commented:
> However, I'm not sure if the debug exporter is the right place to put it in and who need this

I put it there cause then the configuration requirement is minimal. I often log records to debug the broker and noticed that the elasticsearch exporter is helpful to explain and discuss Zeebe. But as setting up elastic is some effort I thought this would be a nice addition. It also kind of provides a way to export the latest records as JSON over the `/records` endpoint, which is nice to share or attach the log to a issue. As it does not add any additional dependencies and is optional I thought it might be okay.

npepinpe commented:
I think the idea is nice, but bundling everything into an exporter is strange; how about having a debug mode which is more than just an exporter, with the exporter being a component of it?

saig0 commented:
I like that it is built as an exporter and that it is in the distro. But I would prefer an extra exporter because the HTTP is not related to the DEBUG one.

menski commented:
I reverted the changes on the existing debug exporter but renamed it to `DebugLogExporter`. I then extracted the `DebugHttpExporter` to an own exporter. The `DebugHttpServer` now serves all resources including javascript and css. The `DebugHttpServer` and therefore the `DebugHttpExporter` will fail if the server cannot be started or a resource is not found.

I also made requested changes to the javascript logic and the cursor is now a pointer to indicate that the table rows can be expanded.

I updated the `zeebe.cfg.toml` to contain both exporters.
